libraries/ms-drivers/src/stm32f0xx/mcp3427_adc.c

```diff
@@ -0,0 +1,180 @@
+#include "mcp3427_adc.h"
+#include "fsm.h"
+#include "log.h"
+#include "mcp3427_adc_defs.h"
+#include "soft_timer.h"
+
+// Note: we always read from both channels before reporting, taking 100ms between reports. If timing
+// is an issue, an optimization for boards (i.e. solar) using only one channel is to only read that
+// channel, so reports take 50ms.
+
+// We use a combination of the address and I2C port in the event data to only transition on events
+// that we raised. Possible optimization if necessary: make prv_get_chip_identifier public, then
+// keep a reverse lookup table of identifier -> MCP3427 storage and only call the appropriate
+// MCP3427 when receiving an MCP3427 event. That way, using n MCP3427s generates n calls to
+// |mcp3427_process_event| per cycle rather than n^2 if we pass every event to every MCP3427.
+// (I think this was done on MSXII.)
+
+#define MCP3427_FSM_NAME "MCP3427 FSM"
+
+FSM_DECLARE_STATE(channel_1_trigger);
+FSM_DECLARE_STATE(channel_1_readback);
+FSM_DECLARE_STATE(channel_2_trigger);
+FSM_DECLARE_STATE(channel_2_readback);
+
+FSM_STATE_TRANSITION(channel_1_trigger) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_ready_event, channel_1_readback);
+}
+
+FSM_STATE_TRANSITION(channel_1_readback) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_trigger_event, channel_2_trigger);
+}
+
+FSM_STATE_TRANSITION(channel_2_trigger) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_ready_event, channel_2_readback);
+}
+
+FSM_STATE_TRANSITION(channel_2_readback) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_trigger_event, channel_1_trigger);
+}
+
+static uint16_t prv_get_chip_identifier(Mcp3427Storage *storage) {
+  // used to gate events we raised to only this MCP3427
+  return (storage->addr << 8) | (storage->port == I2C_PORT_1 ? 0 : 1);
```

Comment:
Does this work if you don't use a ternary operator since it should evaluate to 1 or 0?

Reply:
Eg when I do printf("%d", ((1 << 1) | (1 == 1))); , it gives me 3

Reply:
It will work without a ternary operator, I thought that would be a little less clear though.